File: python/src/transactions.py
from dataclasses import replace


def transact(db, ops):
    working = db
    for op in ops:
        working = op(working)
    final_layer = working.layers[-1]
    return replace(db, layers=db.layers + [final_layer], top_id=working.top_id, curr_time=db.curr_time + 1)


def what_if(db, ops):
    working = db
    for op in ops:
        working = op(working)
    final_layer = working.layers[-1]
    return replace(db, layers=db.layers + [final_layer], top_id=working.top_id)

File: clojure/src/circle_db/transactions.clj
(ns circle-db.transactions)

(defn- apply-ops [db ops]
  (reduce (fn [d op] (op d)) db ops))

(defn transact [db-atom ops]
  (swap! db-atom
         (fn [db]
           (let [working (apply-ops db ops)
                 final-layer (last (:layers working))]
             (assoc db
                    :layers (conj (:layers db) final-layer)
                    :top-id (:top-id working)
                    :curr-time (inc (:curr-time db)))))))

(defn what-if [db ops]
  (let [working (apply-ops db ops)
        final-layer (last (:layers working))]
    (assoc db
           :layers (conj (:layers db) final-layer)
           :top-id (:top-id working))))
